umbraco/Umbraco-CMS
# Configuration to control the creation of default data

Umbraco creates various default data for a new installation - a language, various data types, media types, relation types and member types.  There are scenarios where a user may not want this default content created.  Of course they can just delete it, but if they are working in a team and rather than using a shared database working locally with fresh installs and retrieving schema and content using Umbraco Deploy, uSync or similar, this default content can cause issues.

Examples are:

- Not wanting to use "en-US" as the default language.  With others set up on a Cloud site, a fresh install restoring a new installation from a clone will get the ones from the schema files but still also create "en-US" again.  They'll fight to be the default. and the schema file for this could then get checked in and recreated in the production site.
- Not all data types are useful (e.g. drop down without prevalues) or wanted in all installations.
- Custom media types may be created and preferred to the default ones.

This PR adds some additional configuration to customise this.  By default all default data is created as it currently is, but it you want to, you can add configuration to omit the creation of certain items.

The configuration sits within `Umbraco:Cms` and an example looks like this (note that I've updated this following feedback on the original PR):

```
  "InstallDefaultData": {
    "Languages": {
      "InstallData": "Values",
      "Values": [
        "en-US"
      ]
    },
    "DataTypes": {
      "InstallData": "ExceptValues",
      "Values": [
        "0225af17-b302-49cb-9176-b9f35cab9c17"
      ]
    },
    "MediaTypes": {
      "InstallData": "ExceptValues",
      "Values": [
        "c4b1efcf-a9d5-41c4-9621-e9d273b52a9c"
      ]
    },
    "MemberTypes": {
      "InstallData": "None"
    }
  }
```

Each entity you can control has a pair:

- `CreateDefaultX` - an enum with a value of `None`, `CreateOnly`, `CreateAllExcept` and `All`.
    - `None` means none of the default data of that type that is normally created will be (apart from currently one exception, the string label, as that's used in Umbraco itself).
    - `All` means all are created, and is the default.
    - `CreateOnly` and `CreateAllExcept` means to create only those or all except those respectively, listed in the following setting.
 - `SelectedDefaultX` - a list of Guids, aliases or keys (as appropriate) identifying the items to create or not to create depending on the previous setting.

To Test
- Do a fresh install with various combinations of the above settings and verify the expected default data is created.
